File: app/src/lib/clmm.ts
// Browser CLMM price read — the client twin of the machine's on-chain price seam.
// Mirrors scripts/layouts.ts (pinned PoolState/ObservationState offsets,
// ground-truthed in H6a) and scripts/twap.ts (cumulative-tick TWAP) EXACTLY, the
// same way housemath.ts mirrors crates/house-math. The pinned offsets are the
// ones verify-layouts.ts guards against the live devnet program; keep them
// identical here. Uses the "buffer" polyfill like the rest of the app.
import { PublicKey } from "@solana/web3.js";
import { Buffer } from "buffer";

// ---- devnet CHIP/WSOL demo market (scripts/raydium-constants.ts) ----
export const CLMM_PROGRAM_ID = new PublicKey("DRayAUgENGQBKVaX8owNhgzkEDyoHTGVEGHVJT1E9pfH");
export const CHIP_MINT = new PublicKey("75zyWBYdFSNNFKDaTdEu9nZWdHaZCuuCd7tgCCxi2w6p");
export const WSOL_MINT = new PublicKey("So11111111111111111111111111111111111111112");

// ---- PoolState pinned offsets (span 1544), from scripts/layouts.ts ----
const POOL_SPAN = 1544;
const POOL = { observationId: 201, sqrtPriceX64: 253, tickCurrent: 269 } as const;
// ---- ObservationState pinned offsets (span 4483) ----
const OBS = { SPAN: 4483, observationIndex: 17, observations: 51, ITEM_STRIDE: 44, ITEM_ts: 0, ITEM_cum: 4, COUNT: 100 } as const;

function u128LE(b: Buffer, o: number): bigint { return b.readBigUInt64LE(o) + (b.readBigUInt64LE(o + 8) << 64n); }

export interface PoolView { sqrtPriceX64: bigint; tickCurrent: number; price: number; span: number; }
/** Q64.64 sqrt price → CHIP-per-SOL (mintA = WSOL, mintB = CHIP, equal decimals). */
export function decodePool(b: Buffer): PoolView {
  const sqrtPriceX64 = u128LE(b, POOL.sqrtPriceX64);
  const num = Number(sqrtPriceX64) / 2 ** 64;
  return { sqrtPriceX64, tickCurrent: b.readInt32LE(POOL.tickCurrent), price: num * num, span: b.length };
}

export interface Obs { ts: number; tickCum: bigint; }
/** Initialized observations (ts>0), de-duped, ascending by ts (twap.ts). */
export function collectObservations(b: Buffer): Obs[] {
  const seen = new Map<number, bigint>();
  for (let i = 0; i < OBS.COUNT; i++) {
    const off = OBS.observations + i * OBS.ITEM_STRIDE;
    const ts = b.readUInt32LE(off + OBS.ITEM_ts);
    if (ts > 0) seen.set(ts, b.readBigInt64LE(off + OBS.ITEM_cum));
  }
  return [...seen.entries()].map(([ts, tickCum]) => ({ ts, tickCum })).sort((a, b) => a.ts - b.ts);
}

/** Cumulative tick at time t; extrapolates past the newest obs using currentTick. */
function cumulativeAt(obs: Obs[], currentTick: number, t: number): bigint | null {
  if (obs.length === 0) return null;
  const newest = obs[obs.length - 1];
  if (t >= newest.ts) return newest.tickCum + BigInt(currentTick) * BigInt(t - newest.ts);
  if (t < obs[0].ts) return null;
  for (let i = obs.length - 1; i > 0; i--) {
    const b = obs[i], a = obs[i - 1];
    if (t >= a.ts && t <= b.ts) {
      if (b.ts === a.ts) return a.tickCum;
      const rate = Number(b.tickCum - a.tickCum) / (b.ts - a.ts);
      return a.tickCum + BigInt(Math.round(rate * (t - a.ts)));
    }
  }
  return null;
}

export type TwapStatusKind = "LIVE" | "STALE";
export interface TwapResult {
  status: TwapStatusKind; reason: string;
  avgTick: number | null; price: number | null; // CHIP per SOL
  coverageSecs: number; staleSecs: number; obsCount: number;
}
export function computeTwap(obs: Obs[], currentTick: number, now: number, windowSecs: number, maxStaleness: number): TwapResult {
  const base = { obsCount: obs.length };
  if (obs.length === 0) return { status: "STALE", reason: "cold-start: no observations yet", avgTick: null, price: null, coverageSecs: 0, staleSecs: Infinity, ...base };
  const newest = obs[obs.length - 1], oldest = obs[0];
  const staleSecs = now - newest.ts;
  const coverageSecs = newest.ts - oldest.ts;
  const cumNow = cumulativeAt(obs, currentTick, now)!;
  const cumThen = cumulativeAt(obs, currentTick, now - windowSecs);
  if (cumThen === null) return { status: "STALE", reason: `cold-start: history ${coverageSecs}s < window ${windowSecs}s`, avgTick: null, price: null, coverageSecs, staleSecs, ...base };
  if (staleSecs > maxStaleness) return { status: "STALE", reason: `stale: newest obs ${staleSecs}s old > max ${maxStaleness}s`, avgTick: null, price: null, coverageSecs, staleSecs, ...base };
  const avgTick = Number(cumNow - cumThen) / windowSecs;
  return { status: "LIVE", reason: "ok", avgTick, price: Math.pow(1.0001, avgTick), coverageSecs, staleSecs, ...base };
}

// -------------------- price-status (the machine's commit gate, client-side) --------------------

export type PriceStatusKind = "LIVE" | "UNSTABLE" | "STALE";
export interface PriceStatus {
  kind: PriceStatusKind;          // LIVE = fresh + in band; UNSTABLE = band exceeded; STALE = obs too old / cold
  label: string;                  // "LIVE" | "PRICE UNSTABLE" | "STALE"
  reason: string;                 // human explanation (the on-chain refusal reason)
  spot: number | null;            // CHIP per SOL
  twap: number | null;            // CHIP per SOL (the price_at_commit the machine would snapshot)
  twap1e12: bigint | null;        // TWAP scaled 1e12 (the on-chain fixed point)
  bandBp: number | null;          // |spot − twap| / twap in bp
  staleSecs: number;              // age of the newest observation
  coverageSecs: number; obsCount: number;
  commitAllowed: boolean;         // would spin_commit_dual accept right now?
}

/** Classify the pool the way eval_price_gates does: staleness first (STALE), then
 * band (PRICE UNSTABLE). `now` is cluster unix time. windowSecs/maxStaleness/bandBp
 * come from the DualMachine params (identical gate as the on-chain seam). */
export function priceStatus(
  poolData: Buffer, obsData: Buffer, now: number,
  windowSecs: number, maxStalenessSecs: number, bandBp: number,
): PriceStatus {
  const pool = decodePool(poolData);
  const okSpan = pool.span >= POOL_SPAN && obsData.length >= OBS.SPAN;
  const obs = collectObservations(obsData);
  const twap = computeTwap(obs, pool.tickCurrent, now, windowSecs, maxStalenessSecs);
  const spot = pool.price;
  if (!okSpan) {
    return { kind: "STALE", label: "STALE", reason: "pool/observation account layout unexpected", spot, twap: null, twap1e12: null, bandBp: null, staleSecs: twap.staleSecs, coverageSecs: twap.coverageSecs, obsCount: twap.obsCount, commitAllowed: false };
  }
  if (twap.status === "STALE") {
    return { kind: "STALE", label: "STALE", reason: twap.reason, spot, twap: null, twap1e12: null, bandBp: null, staleSecs: twap.staleSecs, coverageSecs: twap.coverageSecs, obsCount: twap.obsCount, commitAllowed: false };
  }
  const twapPrice = twap.price!;
  const band = Math.round((Math.abs(spot - twapPrice) / twapPrice) * 10_000);
  const twap1e12 = BigInt(Math.round(twapPrice * 1e12));
  if (band > bandBp) {
    return { kind: "UNSTABLE", label: "PRICE UNSTABLE", reason: `spot ${spot.toFixed(2)} drifted ${band}bp from TWAP ${twapPrice.toFixed(2)} (> ${bandBp}bp band)`, spot, twap: twapPrice, twap1e12, bandBp: band, staleSecs: twap.staleSecs, coverageSecs: twap.coverageSecs, obsCount: twap.obsCount, commitAllowed: false };
  }
  return { kind: "LIVE", label: "LIVE", reason: "fresh and in band", spot, twap: twapPrice, twap1e12, bandBp: band, staleSecs: twap.staleSecs, coverageSecs: twap.coverageSecs, obsCount: twap.obsCount, commitAllowed: true };
}

File: crates/house-math/src/clmm.rs
//! On-chain Raydium CLMM price reader (H6b-3). Parses `PoolState` (spot, via
//! sqrt_price) and `ObservationState` (cumulative-tick TWAP) straight from raw
//! account bytes using the H6a PINNED offsets — the ones the `verify-layouts.ts`
//! regression guard checks against the live devnet program, NOT the published
//! structs. Reuses [`crate::price`] and [`crate::twap`]; no math is re-derived.
//!
//! Orientation assumption (the H6a demo pool): mintA = WSOL (9 dec, quote),
//! mintB = the machine's token, so `price = mintB per mintA = token per SOL`, and
//! with 9-dec tokens the raw ratio already equals whole-token-per-whole-SOL. The
//! caller owner-checks the accounts and matches the machine's stored pubkeys.
//!
//! The TWAP reproduces `scripts/twap.ts` exactly (verified in `ground_truth`):
//! `cum(now)` extrapolates the newest observation at the current tick, and
//! `cum(now−window)` interpolates the ring — both in integer arithmetic.

use crate::price::price_1e12_from_sqrt_x64;
use crate::twap::{avg_tick, StaleReason, TwapRead};

// ---- pinned PoolState offsets (span 1544) ----
pub const POOL_SPAN: usize = 1544;
pub const POOL_MINT_A: usize = 73;
pub const POOL_MINT_B: usize = 105;
pub const POOL_OBSERVATION_ID: usize = 201;
pub const POOL_SQRT_PRICE_X64: usize = 253;
pub const POOL_TICK_CURRENT: usize = 269;
// ---- pinned ObservationState offsets (span 4483) ----
pub const OBS_SPAN: usize = 4483;
pub const OBS_INDEX: usize = 17;
pub const OBS_POOL_ID: usize = 19;
pub const OBS_OBSERVATIONS: usize = 51;
pub const OBS_ITEM_STRIDE: usize = 44;
pub const OBS_ITEM_TS: usize = 0; // u32, relative to item start
pub const OBS_ITEM_CUM: usize = 4; // i64, relative to item start
pub const OBS_COUNT: usize = 100;

#[cfg(test)]
fn u16_le(b: &[u8], o: usize) -> u16 { u16::from_le_bytes([b[o], b[o + 1]]) }
fn u32_le(b: &[u8], o: usize) -> u32 { let mut a = [0u8; 4]; a.copy_from_slice(&b[o..o + 4]); u32::from_le_bytes(a) }
fn i32_le(b: &[u8], o: usize) -> i32 { let mut a = [0u8; 4]; a.copy_from_slice(&b[o..o + 4]); i32::from_le_bytes(a) }
fn i64_le(b: &[u8], o: usize) -> i64 { let mut a = [0u8; 8]; a.copy_from_slice(&b[o..o + 8]); i64::from_le_bytes(a) }
fn u128_le(b: &[u8], o: usize) -> u128 { let mut a = [0u8; 16]; a.copy_from_slice(&b[o..o + 16]); u128::from_le_bytes(a) }

/// PoolState.sqrt_price (Q64.64), pinned offset 253.
pub fn pool_sqrt_price_x64(pool: &[u8]) -> u128 { u128_le(pool, POOL_SQRT_PRICE_X64) }
/// PoolState.tick_current (i32), pinned offset 269.
pub fn pool_tick_current(pool: &[u8]) -> i32 { i32_le(pool, POOL_TICK_CURRENT) }
/// Spot price scaled 1e12 from PoolState.sqrt_price (equal-decimals orientation).
pub fn pool_spot_1e12(pool: &[u8]) -> u128 { price_1e12_from_sqrt_x64(pool_sqrt_price_x64(pool)) }
/// PoolState.observation_id / ObservationState.pool_id — 32-byte pubkeys, for the
/// caller's cross-link checks (pool ↔ observation).
pub fn pool_observation_id(pool: &[u8]) -> [u8; 32] { let mut k = [0u8; 32]; k.copy_from_slice(&pool[POOL_OBSERVATION_ID..POOL_OBSERVATION_ID + 32]); k }
pub fn obs_pool_id(obs: &[u8]) -> [u8; 32] { let mut k = [0u8; 32]; k.copy_from_slice(&obs[OBS_POOL_ID..OBS_POOL_ID + 32]); k }

fn obs_item(obs: &[u8], i: usize) -> (u32, i64) {
    let off = OBS_OBSERVATIONS + i * OBS_ITEM_STRIDE;
    (u32_le(obs, off + OBS_ITEM_TS), i64_le(obs, off + OBS_ITEM_CUM))
}

/// The parsed CLMM price reading. `twap` is the raw TWAP outcome (staleness is
/// the caller's gate, via `newest_obs_age_secs`); `NotReady` means the ring
/// doesn't cover the window (cold-start).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ClmmReading {
    pub spot_1e12: u128,
    pub twap: TwapRead,
    pub newest_obs_age_secs: u32,
    pub obs_count: u32,
    pub coverage_secs: u32,
}

/// Parse spot + TWAP from raw PoolState/ObservationState bytes. `now` is the
/// current unix time (u32, matching the observation timestamps); `window_secs`
/// is the TWAP window. Returns `None` only on a length/shape failure (the caller
/// has already owner-checked the accounts). Reproduces scripts/twap.ts exactly.
pub fn read_clmm_price(pool: &[u8], obs: &[u8], now: u32, window_secs: u32) -> Option<ClmmReading> {
    if pool.len() < POOL_SPAN || obs.len() < OBS_SPAN {
        return None;
    }
    let spot_1e12 = pool_spot_1e12(pool);
    let cur_tick = pool_tick_current(pool);
    let target = now.wrapping_sub(window_secs); // now ≫ window for real timestamps

    // one scan: newest (max ts), oldest (min ts), and the bracket a≤target≤b.
    let mut newest_ts = 0u32;
    let mut newest_cum = 0i64;
    let mut oldest_ts = u32::MAX;
    let mut count = 0u32;
    let (mut a_ts, mut a_cum, mut a_found) = (0u32, 0i64, false);
    let (mut b_ts, mut b_cum, mut b_found) = (u32::MAX, 0i64, false);
    for i in 0..OBS_COUNT {
        let (ts, cum) = obs_item(obs, i);
        if ts == 0 {
            continue; // uninitialized slot
        }
        count += 1;
        if ts > newest_ts { newest_ts = ts; newest_cum = cum; }
        if ts < oldest_ts { oldest_ts = ts; }
        if ts <= target && (!a_found || ts > a_ts) { a_ts = ts; a_cum = cum; a_found = true; }
        if ts >= target && (!b_found || ts < b_ts) { b_ts = ts; b_cum = cum; b_found = true; }
    }
    if count == 0 {
        return Some(ClmmReading { spot_1e12, twap: TwapRead::NotReady(StaleReason::WindowNotCovered), newest_obs_age_secs: u32::MAX, obs_count: 0, coverage_secs: 0 });
    }
    let age = now.wrapping_sub(newest_ts);
    let coverage = newest_ts.wrapping_sub(oldest_ts);

    // cum(now): extrapolate the newest observation at the current tick.
    let cum_now = if now >= newest_ts {
        newest_cum.wrapping_add(cur_tick as i64 * (now - newest_ts) as i64)
    } else {
        newest_cum
    };
    // cum(now − window): extrapolate if target is past the newest, else interpolate the ring.
    let cum_then: Option<i64> = if target >= newest_ts {
        Some(newest_cum.wrapping_add(cur_tick as i64 * (target - newest_ts) as i64))
    } else if !a_found {
        None // target older than the oldest observation → window not covered
    } else if !b_found || a_ts == b_ts {
        Some(a_cum)
    } else {
        // integer interpolation, matching scripts/twap.ts::cumulativeAt to the unit
        Some(a_cum + (b_cum - a_cum) * (target - a_ts) as i64 / (b_ts - a_ts) as i64)
    };

    let twap = match cum_then {
        Some(ct) => match avg_tick(ct, target, cum_now, now) {
            Some(a) => TwapRead::Live { avg_tick: a },
            None => TwapRead::NotReady(StaleReason::WindowNotCovered),
        },
        None => TwapRead::NotReady(StaleReason::WindowNotCovered),
    };
    Some(ClmmReading { spot_1e12, twap, newest_obs_age_secs: age, obs_count: count, coverage_secs: coverage })
}

#[cfg(test)]
mod proofs {
    use super::*;
    use crate::price::price_1e12_at_tick;

    const POOL_BYTES: &[u8] = include_bytes!("../fixtures/clmm_pool.bin");
    const OBS_BYTES: &[u8] = include_bytes!("../fixtures/clmm_obs.bin");

    /// PROOF (ground truth): the parser fed CAPTURED LIVE devnet bytes reads the
    /// pinned offsets correctly and reproduces the H6a twap-status computation
    /// exactly. Reference values pinned from scripts/twap.ts + layouts.ts against
    /// the same bytes (pool 9n6LAVick…, observation 7nPBDXZ…, captured 2026-07-06):
    ///   sqrt_price_x64 575415330658078091153, tick 68807, spot 973.0237 CHIP/SOL,
    ///   observation index 11, newest ts 1783301680 / cum 44286166, 12 obs, 645s
    ///   coverage; TWAP over 300s @ now=newest → avg_tick 68517.
    #[test]
    fn ground_truth_matches_h6a() {
        assert_eq!(POOL_BYTES.len(), POOL_SPAN, "pool fixture span");
        assert_eq!(OBS_BYTES.len(), OBS_SPAN, "obs fixture span");

        // spot: bit-exact against the pinned integer (== round(973.0236854537752e12)).
        assert_eq!(pool_sqrt_price_x64(POOL_BYTES), 575415330658078091153);
        assert_eq!(pool_tick_current(POOL_BYTES), 68807);
        assert_eq!(pool_spot_1e12(POOL_BYTES), 973023685453775, "spot_1e12 == H6a to the unit");

        // the observation offsets: index and the newest + a couple of ring entries
        // (from the H6a-captured cumulative-tick series).
        assert_eq!(u16_le(OBS_BYTES, OBS_INDEX), 11, "observationIndex @17");
        assert_eq!(obs_item(OBS_BYTES, 11), (1783301680, 44286166), "newest observation @ index 11");
        assert_eq!(obs_item(OBS_BYTES, 0), (1783301035, 0), "first observation (cold seed)");
        assert_eq!(obs_item(OBS_BYTES, 1), (1783301056, 1447152), "second observation");
        // the pool↔observation cross-link fields are present (32-byte pubkeys, the
        // on-chain reader require_keys_eq's them against the machine's stored ids).
        assert_ne!(pool_observation_id(POOL_BYTES), [0u8; 32], "pool stores an observation id");
        assert_ne!(obs_pool_id(OBS_BYTES), [0u8; 32], "observation stores a pool id");
    }

    /// PROOF: the TWAP over the ring reproduces H6a exactly at a fresh `now`.
    #[test]
    fn ground_truth_twap_matches_h6a() {
        // now = the newest observation's ts → fresh; window 300s (H6a demo window).
        let r = read_clmm_price(POOL_BYTES, OBS_BYTES, 1783301680, 300).unwrap();
        assert_eq!(r.obs_count, 12);
        assert_eq!(r.coverage_secs, 645);
        assert_eq!(r.newest_obs_age_secs, 0);
        match r.twap {
            TwapRead::Live { avg_tick } => {
                // integer avg_tick == scripts/twap.ts (its float 68517.70 floors here).
                assert_eq!(avg_tick, 68517, "TWAP avg_tick matches H6a");
                // and the fixed-point TWAP price is ~945 CHIP/SOL (H6a float 945.23).
                let twap_1e12 = price_1e12_at_tick(avg_tick as i32);
                assert!((945_000_000_000_000..=946_000_000_000_000).contains(&twap_1e12),
                    "twap_1e12 ~945e12: {twap_1e12}");
            }
            other => panic!("expected LIVE twap, got {other:?}"),
        }
    }

    /// PROOF: a `now` far past the newest observation (the live-STALE case: the
    /// keeper stopped) yields a huge age, which the caller's staleness gate uses
    /// to refuse — the reader itself never panics.
    #[test]
    fn stale_now_reports_large_age() {
        let r = read_clmm_price(POOL_BYTES, OBS_BYTES, 1783301680 + 20_000, 300).unwrap();
        assert!(r.newest_obs_age_secs >= 20_000, "age reflects the quiet pool");
    }

    /// PROOF: a cold ring (no initialized observations) is NotReady with a
    /// forced-huge age, so the caller refuses; spot is still readable.
    #[test]
    fn cold_ring_is_not_ready() {
        let mut obs = OBS_BYTES.to_vec();
        for i in 0..OBS_COUNT {
            let off = OBS_OBSERVATIONS + i * OBS_ITEM_STRIDE;
            for b in obs.iter_mut().skip(off).take(4) { *b = 0; } // zero every blockTimestamp
        }
        let r = read_clmm_price(POOL_BYTES, &obs, 1783301680, 300).unwrap();
        assert_eq!(r.twap, TwapRead::NotReady(StaleReason::WindowNotCovered));
        assert_eq!(r.newest_obs_age_secs, u32::MAX);
        assert_eq!(r.spot_1e12, 973023685453775, "spot still readable when cold");
    }
}
